keps/sig-scheduling/20190312-pod-affinity-and-anti-affinity-support-gt-and-lt-operators.md

```diff
@@ -0,0 +1,193 @@
+      - labelSelector:
+          matchExpressions:
+          - key: foo
+            operator: In
+            values:
+            - 20
+            - 21
+            - 22
+            - 23
+            ...
+            - 28
+            - 29
+            - 30
+        topologyKey: failure-domain.beta.kubernetes.io/zone
+```
+
+ This is not an ideal solution. A promising solution is to provide users with `Gt` and `Lt` 
+ operators, giving users the ability to specify the scope of the tag. Users can achieve 
+ this in this way:
+ ```yaml
+spec:
+  affinity:
+    podAffinity:
+      requiredDuringSchedulingIgnoredDuringExecution:
+      - labelSelector:
+          matchExpressions:
+          - key: foo
+            operator: Gt
+            values:
+            - 20
+          - key: foo
+            operator: Lt
+            values:
+            - 30
+        topologyKey: failure-domain.beta.kubernetes.io/zone
+```
+
+ ### Risks and Mitigations
+
+ Along with this feature, the biggest risk should be performance. This may also be the reason 
+ why the `Gt` and `Lt` operators are not supported when Pod affinity/anti-affinity is 
+ first proposed. In order to understand the impact of these changes, we need to understand the their performance implications. We will perform performance/benchmark tests on this change and backward compatibility test.
+
+ Compared to the time that inter-pod affinity was introduced, performance of the scheduler has been greatly improved(https://github.com/kubernetes/kubernetes/pull/74041#issuecomment-466191359), and the processing of Pod affinity/anti-affinity has been surprisingly Optimized(https://github.com/kubernetes/kubernetes/pull/67788).
+
+ ## Design Details
+ ### Content
+  We will abstract the `Node` and `Pod` label selection feature, and finally the `NodeSelector` and `PodSelector` will be based on the `LabelSelectorRequirement` implementation:
+
+ API of `NodeSelector` is defined as below:
+ ```go
+type NodeSelector struct {
+  NodeSelectorTerms []NodeSelectorTerm
+}
+
+type NodeSelectorTerm struct {
+  MatchExpressions []LabelSelectorRequirement
+  MatchFields []LabelSelectorRequirement
+}
+```
+
+ API of `PodSelector` is defined as below:
+ ```go
+type PodSelector struct {
+  MatchLabels map[string]string
+  MatchExpressions []LabelSelectorRequirement
+}
+```
+
+ API of `LabelSelectorRequirement` is defined as below:
+
+ ```go
+type LabelSelectorRequirement struct {
+  Key string
+  Operator LabelSelectorOperator
+  Values []string
+}
+
+type LabelSelectorOperator string
+
+const (
+  LabelSelectorOpIn           LabelSelectorOperator = "In"
+  LabelSelectorOpNotIn        LabelSelectorOperator = "NotIn"
+  LabelSelectorOpExists       LabelSelectorOperator = "Exists"
+  LabelSelectorOpDoesNotExist LabelSelectorOperator = "DoesNotExist"
+  LabelSelectorOpGt           LabelSelectorOperator = "Gt"
+  LabelSelectorOpLt           LabelSelectorOperator = "Lt"
+)
+```
```

Comment:
Last thought: whatever design you come up with has to maintain backwards compatibility. Existing pods must continue to work the same without any changes.

Reply:
Yes, this has also clarified in the past that the definition of Pod will not be changed.

Reply:
I don't understand that response, this proposal definitely looks like it is modifying types which are part of a Pod?

Reply:
Backward compatibility is a requirement for this change. The intention of this KEP is that existing pod specs should work without needing any changes.